### PR TITLE
[Preprocessing] Add a one-off pattern to fuse attention with transpose.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
The attention ops in SDXL models are usually followed by a `tensor.expand_shape` and a `transpose`. It is more natural to fold these in with the attention for codegeneration. This is done as a one-off pattern for now. Ideally the attention ops can be fused with any of its elementwise consumers when attention is handled natively by the backend pass-pipelines.
More details are in https://github.com/iree-org/iree/issues/17673.

### CODE REVIEW

**Comment by IanWood1** [comment]
Great, I didn't realize the attention changes landed. I'll get back to https://github.com/iree-org/iree/issues/17673

File: build_tools/pkgci/external_test_suite/attention_and_matmul_spec.mlir
Comment (raikonenfnu) [inline build_tools/pkgci/external_test_suite/attention_and_matmul_spec.mlir:1]: Seems like test is passing, but if we just replace the default attn_matmul_spec to be the transposed one, wouldn't the rest of the tests with regular/non transposed-expanded attention fail?

Comment (ScottTodd) [reply build_tools/pkgci/external_test_suite/attention_and_matmul_spec.mlir:1]: As of https://github.com/iree-org/iree/commit/44808e143533e694644da13b75a8f8358ee289cf, we're using only that spec file in two specific tests:

https://github.com/iree-org/iree/blob/f89a7da8b8671c20ea8a0f87a416de49b087b0eb/experimental/regression_suite/shark-test-suite-models/sd3/test_mmdit.py#L87 https://github.com/iree-org/iree/blob/f89a7da8b8671c20ea8a0f87a416de49b087b0eb/experimental/regression_suite/shark-test-suite-models/sdxl/test_unet.py#L93

Prior to that commit we were using it for all pytorch models (oops ;P)

File: compiler/src/iree/compiler/GlobalOptimization/RaiseSpecialOps.cpp [hunk 1]
```diff
@@ -990,6 +990,170 @@ class ConcatenateNegateAndSlicePattern
...
+          transposeLikeOp, "expand shape user is not a transpose");
+    }
+
+    // Check attention op indexing maps.
+    AffineExpr d0, d1, d2, d3, d4, d5;
+    bindDims(rewriter.getContext(), d0, d1, d2, d3, d4, d5);
+    auto getIndexingMap = [&](int n, ArrayRef<AffineExpr> results) {
+      return AffineMap::get(n, 0, results, rewriter.getContext());
+    };
+    SmallVector<AffineMap> expectedMaps = {
```
Comment (raikonenfnu) [inline compiler/src/iree/compiler/GlobalOptimization/RaiseSpecialOps.cpp]: [No action required / side note] Maybe we should also add a helper to get standard attention map to help for future pattern matches. :)

File: compiler/src/iree/compiler/GlobalOptimization/RaiseSpecialOps.cpp [hunk 2]
```diff
@@ -990,6 +990,170 @@ class ConcatenateNegateAndSlicePattern
...
+    if (!expandShapeOp->hasOneUse()) {
+      return rewriter.notifyMatchFailure(attentionOp,
+                                         "expand shape op has multiple uses");
+    }
+    auto transposeLikeOp =
+        dyn_cast<linalg::LinalgOp>(*expandShapeOp->user_begin());
+    if (!transposeLikeOp) {
+      return failure();
+    }
+    if (!(transposeLikeOp.getNumDpsInputs() == 1 &&
```
Comment (raikonenfnu) [inline compiler/src/iree/compiler/GlobalOptimization/RaiseSpecialOps.cpp]: [No action required] I'm surprised, we haven't got a helper function for the transposeLikeOp either. I know we have `LinalgOpInfo(op).isTranspose()` but I don't think it's doing this haha.

**Review by raikonenfnu — APPROVED** [review]
LGTM!

File: build_tools/pkgci/external_test_suite/attention_and_matmul_spec.mlir
Comment (raikonenfnu) [reply build_tools/pkgci/external_test_suite/attention_and_matmul_spec.mlir:1]: Should probably put warning/note somewhere, so we wouldn't psych our future selves hahaha

Comment (ScottTodd) [reply build_tools/pkgci/external_test_suite/attention_and_matmul_spec.mlir:1]: fyi @saienduri

File: compiler/src/iree/compiler/GlobalOptimization/RaiseSpecialOps.cpp (hunk 1 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/GlobalOptimization/RaiseSpecialOps.cpp]: See https://github.com/iree-org/iree/issues/17673

**Comment by MaheshRavishankar** [comment]
@raikonenfnu  I gave up trying to make this work by default everywhere. So instead moved it to preprocessing. Wherever we use the new attention script, we add this preprocessing flag.

**Pull request merged**